### PR TITLE
[WIP] Create the Persistence component

### PR DESCRIPTION
**Pull request by tgalopin**
This PR implements the Persistence library started in https://github.com/acmephp/persistence/pull/1.

It is a work in progress as the OOP structure is still not fully stable and may change in the near future.

### CODE REVIEW

File: src/Persistence/Serializer/PemNormalizer.php [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    {
+        return new $class($data);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function supportsNormalization($data, $format = null)
+    {
+        return is_object($data) && ($data instanceof Certificate || $data instanceof Key);
```
Comment (jderusse) [inline src/Persistence/Serializer/PemNormalizer.php]: Better rely on interface

File: src/Persistence/Formatter/DefaultFormatter.php [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+        ];
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function createDomainDistinguishedNameFiles($domain, DistinguishedName $distinguishedName)
+    {
+        return [
+            'certs/'.$domain.'/certificate_request.json' => $this->serializeJson($distinguishedName),
```
Comment (jderusse) [inline src/Persistence/Formatter/DefaultFormatter.php]: file's name don't match his content

File: src/Persistence/Formatter/DefaultFormatter.php [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+        ];
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function createDomainCertificateFiles($domain, KeyPair $keyPair, Certificate $certificate)
```
Comment (jderusse) [inline src/Persistence/Formatter/DefaultFormatter.php]: in fact this method don't create file

File: src/Persistence/Formatter/NginxProxyFormatter.php [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+    /**
+     * {@inheritdoc}
+     */
+    public function createDomainCertificateFiles($domain, KeyPair $keyPair, Certificate $certificate)
+    {
+        $fullChain = [];
+
+        while (null !== $certificate) {
+            $fullChain[] = $this->serializePem($certificate);
```
Comment (jderusse) [inline src/Persistence/Formatter/NginxProxyFormatter.php]: This code is partially a dupplicate from the DefaultFormatter

File: src/Persistence/Formatter/AbstractFormatter.php [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+use AcmePhp\Persistence\Serializer\PemNormalizer;
+use Symfony\Component\Serializer\Encoder\JsonEncoder;
+use Symfony\Component\Serializer\Normalizer\GetSetMethodNormalizer;
+use Symfony\Component\Serializer\Serializer;
+use Symfony\Component\Serializer\SerializerInterface;
+
+/**
+ * @author Titouan Galopin <galopintitouan@gmail.com>
+ */
+abstract class AbstractFormatter implements FormatterInterface
```
Comment (jderusse) [inline src/Persistence/Formatter/AbstractFormatter.php]: use composition over inheritance

File: src/Persistence/Loader/DefaultLoader.php [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+
+/**
+ * @author Titouan Galopin <galopintitouan@gmail.com>
+ */
+class DefaultLoader extends AbstractLoader
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function loadAccountKeyPair(AdapterInterface $adapter)
```
Comment (jderusse) [inline src/Persistence/Loader/DefaultLoader.php]: A service shouldn't be an method's argument

File: src/Persistence/Loader/DefaultLoader.php [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+    {
+        return new KeyPair(
+            $this->deserializePem($adapter->read('private/_account/public.pem'), PublicKey::class),
```
Comment (jderusse) [inline src/Persistence/Loader/DefaultLoader.php]: file's path are dupplicated between Formatters and Loaders

**Comment by tgalopin** [comment]
For the moment, this library tries to do too much IMO. We will work on it in a future release, but for now I close this and I will put the persistence layer in the CLI.

**Comment by kevinquinnyo** [comment]
@tgalopin Just my two cents here -- I think that's a good move.  The reason is because I think the big draw to your library for people like me is that I can choose how I persist the Challenge and Certificate objects.  For instance, I'm saving them into a database as part of a CakePHP app.  So saving this data as files is really not important at all to me.  I end up doing that downstream via configuration management tools.

For the CLI utility, however, I can see having the ability to persist the data easily as files might be a nice feature.  But for a lot of use-cases (like mine) we're happy to just have the objects, and decide how to save them.

**Pull request closed without merging**